app/lib/shared/utils.dart

```diff
@@ -361,3 +362,38 @@ bool fixedTimeEquals(String a, String b) {
   }
   return result == 0;
 }
+
+extension StreamBoundedForEach<T> on Stream<T> {
+  /// Executes items using [eachFn] with maximum [concurrency].
+  Future<void> boundedForEach(
+    int concurrency,
+    FutureOr<void> Function(T) eachFn,
+  ) async {
+    final pool = Pool(concurrency);
+    final futures = Queue<Future>();
+    // keeps the size of the futures limited, allows some difference in the run times
+    final maxQueueSize = concurrency * 2;
+    try {
+      await for (final item in this) {
+        final f = pool.withResource(() => eachFn(item));
+        futures.add(f);
+        while (futures.length > maxQueueSize) {
+          await futures.removeFirst();
+        }
```

Comment:
I don't understand this.

Reading: https://pub.dev/documentation/pool/latest/pool/Pool/close.html

I think it's sufficient to just do `await pool.close()`.

```dart
    final pool = Pool(concurrency);
    try {
      await for (final item in this) {
        unawaited(pool.withResource(() async {
          await eachFn(item);
        }));
      }
    } finally {
      await pool.close();
    }
```

maybe actually want to do:

```
    Object? err;
    StackTrace? stackTrace;
    final pool = Pool(concurrency);
    try {
      await for (final item in this) {
        unawaited(pool.withResource(() async {
          try {
            await eachFn(item);
          } catch (e, st) {
            // Take the first error
            if (err != null) {
              err = e;
              stackTrace = st;
            }
          }
        }));
      }
    } finally {
      await pool.close();
      if (err != null) {
        rethrow err, st; // if something like this is possible.
      }
    }
```